cpp/include/cudf/hashing/detail/murmurhash3_x64_128.cuh

```diff
@@ -15,211 +15,96 @@
-  }
-
-  result_type __device__ compute_bytes(std::byte const* data, cudf::size_type const len) const
+  __device__ constexpr result_type compute(T const& key) const
   {
-    auto const nblocks = len / BLOCK_SIZE;
-    uint64_t h1        = m_seed;
-    uint64_t h2        = m_seed;
-
-    // Process all four-byte chunks.
-    for (cudf::size_type i = 0; i < nblocks; i++) {
-      uint64_t k1 = getblock64(data, (i * BLOCK_SIZE));                     // 1st 8 bytes
-      uint64_t k2 = getblock64(data, (i * BLOCK_SIZE) + (BLOCK_SIZE / 2));  // 2nd 8 bytes
-
-      k1 *= c1;
-      k1 = rotate_bits_left(k1, 31);
-      k1 *= c2;
-
-      h1 ^= k1;
-      h1 = rotate_bits_left(h1, 27);
-      h1 += h2;
-      h1 = h1 * 5 + 0x52dce729;
-
-      k2 *= c2;
-      k2 = rotate_bits_left(k2, 33);
-      k2 *= c1;
-
-      h2 ^= k2;
-      h2 = rotate_bits_left(h2, 31);
-      h2 += h1;
-      h2 = h2 * 5 + 0x38495ab5;
-    }
-
-    thrust::tie(h1, h2) = compute_remaining_bytes(data, len, nblocks * BLOCK_SIZE, {h1, h2});
-
-    // Finalize hash.
-    h1 ^= len;
-    h2 ^= len;
-
-    h1 += h2;
-    h2 += h1;
-
-    h1 = fmix64(h1);
-    h2 = fmix64(h2);
-
-    h1 += h2;
-    h2 += h1;
-
-    return {h1, h2};
+    return this->compute_bytes(reinterpret_cast<cuda::std::byte const*>(&key), sizeof(T));
   }
 
- private:
-  uint64_t m_seed{};
-  static constexpr uint32_t BLOCK_SIZE = 16;  // 2 x 64-bit = 16 bytes
-
-  static constexpr uint64_t c1 = 0x87c37b91114253d5UL;
-  static constexpr uint64_t c2 = 0x4cf5ad432745937fUL;
+  cuco::murmurhash3_x64_128<Key> _impl;
 };
 
 template <>
 MurmurHash3_x64_128<bool>::result_type __device__ inline MurmurHash3_x64_128<bool>::operator()(
   bool const& key) const
 {
-  return compute<uint8_t>(key);
+  return this->compute<uint8_t>(key);
 }
 
 template <>
 MurmurHash3_x64_128<float>::result_type __device__ inline MurmurHash3_x64_128<float>::operator()(
   float const& key) const
 {
-  return compute(normalize_nans(key));
+  return this->compute(normalize_nans(key));
 }
 
 template <>
 MurmurHash3_x64_128<double>::result_type __device__ inline MurmurHash3_x64_128<double>::operator()(
   double const& key) const
 {
-  return compute(normalize_nans(key));
+  return this->compute(normalize_nans(key));
 }
 
 template <>
 MurmurHash3_x64_128<cudf::string_view>::result_type
   __device__ inline MurmurHash3_x64_128<cudf::string_view>::operator()(
     cudf::string_view const& key) const
 {
-  auto const data = reinterpret_cast<std::byte const*>(key.data());
-  auto const len  = key.size_bytes();
-  return compute_bytes(data, len);
+  return this->compute_bytes(reinterpret_cast<cuda::std::byte const*>(key.data()),
+                             key.size_bytes());
 }
 
 template <>
 MurmurHash3_x64_128<numeric::decimal32>::result_type
   __device__ inline MurmurHash3_x64_128<numeric::decimal32>::operator()(
     numeric::decimal32 const& key) const
 {
-  return compute(key.value());
+  return this->compute(key.value());
```

Comment:
I've a silly question:  Why does `this` need to be explicitly mentioned here?

Reply:
I personally find that using `this` makes the intention to invoke the private member function `compute` clearer.